File: CheckIn.Api.Common/Models/Create/TaskCreateModel.cs
using CheckIn.Api.Common.Enums;

namespace CheckIn.Api.Common.Models.Create;

public record TaskCreateModel
{
    public required string Title { get; init; }
    public string? Notes { get; init; }
    public DateTime DeadLine { get; init; }


    public required SubtaskMode SubtaskMode { get; init; }
    public List<SubtaskTemplateCreateModel> Subtasks { get; init; } = new();
    public bool RequiresAuthenticationToComplete { get; init; } = true;
    public bool SendInvitesImmediately { get; init; } = false;

    public List<string> InvitedEmails { get; set; } = new();
}

File: checkin_frontend/lib/core/models/task/task_create_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../enums/task_enums.dart';
import '../subtask_template/subtask_template_create_model.dart';

part 'task_create_model.freezed.dart';
part 'task_create_model.g.dart';

@freezed
sealed class TaskCreateModel with _$TaskCreateModel {
  const factory TaskCreateModel({
    required String title,
    String? notes,
    required DateTime deadLine,
    required SubtaskMode subtaskMode,
    @Default(true) bool requiresAuthenticationToComplete,
    @Default([]) List<SubtaskTemplateCreateModel> subtasks,
    @Default([]) List<String> invitedEmails,
    @Default(false) bool sendInvitesImmediately,
    String? allowedDomain,

    @JsonKey(includeToJson: false)
    @Default(null) bool? isDomainValid, // null = unchecked, true = ok, false = suspicious
  }) = _TaskCreateModel;

  factory TaskCreateModel.fromJson(Map<String, dynamic> json) =>
      _$TaskCreateModelFromJson(json);
}